Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/TestCertificateRepository.kt

```diff
@@ -0,0 +1,329 @@
+            .onEach {
+                Timber.tag(TAG).v("TestCertificateContainer data changed: %s", it)
+                storage.testCertificates = it.values.toSet()
+            }
+            .catch {
+                it.reportProblem(TAG, "Failed to snapshot TestCertificateContainer data to storage.")
+                throw it
+            }
+            .launchIn(appScope + dispatcherProvider.Default)
+    }
+
+    suspend fun requestCertificate(test: CoronaTest): TestCertificateContainer {
+        Timber.tag(TAG).d("createDccForTest(test.identifier=%s)", test.identifier)
+
+        val newData = internalData.updateBlocking {
+            if (values.any { it.registrationToken == test.registrationToken }) {
+                Timber.tag(TAG).e("Certificate entry already exists for %s", test.identifier)
+                throw IllegalArgumentException("A certificate was already created for this ${test.identifier}")
+            }
+
+            // TODO do we need additional validation here? last chance to abort?
+
+            val identifier = UUID.randomUUID().toString()
+
+            val certificate = when (test.type) {
+                CoronaTest.Type.PCR -> PCRCertificateContainer(
+                    identifier = identifier,
+                    registeredAt = test.registeredAt,
+                    registrationToken = test.registrationToken,
+                )
+                CoronaTest.Type.RAPID_ANTIGEN -> RACertificateContainer(
+                    identifier = identifier,
+                    registeredAt = test.registeredAt,
+                    registrationToken = test.registrationToken,
+                )
+            }
+            Timber.tag(TAG).d("Adding test certificate entry: %s", certificate)
+            mutate { this[certificate.identifier] = certificate }
+        }
+
+        return newData.values.single { it.registrationToken == test.registrationToken }
+    }
+
+    data class RefreshResult(
+        val certificate: TestCertificateContainer,
+        val error: Exception? = null,
+    )
+
+    suspend fun refresh(identifier: TestCertificateIdentifier? = null): Set<RefreshResult> {
+        Timber.tag(TAG).d("refresh(identifier=%s)", identifier)
+
+        val refreshResults = mutableSetOf<RefreshResult>()
+
+        val workedOnIds = mutableSetOf<TestCertificateIdentifier>()
+
+        internalData.updateBlocking {
+            val toRefresh = values
+                .filter { it.identifier == identifier || identifier == null }
+                .filter { it.isPending }
+
+            mutate {
+                toRefresh.forEach {
+                    workedOnIds.add(it.identifier)
+                    this[it.identifier] = when (it.type) {
+                        CoronaTest.Type.PCR -> (it as PCRCertificateContainer).copy(isUpdatingData = true)
+                        CoronaTest.Type.RAPID_ANTIGEN -> (it as RACertificateContainer).copy(isUpdatingData = true)
```

Comment:
when (it) {
                        is PCRCertificateContainer -> it.copy()
                        is RACertificateContainer -> it.copy()
                    }